module/zfs/zfs_vnops.c

```
@@ -979,13 +981,497 @@ zfs_get_done(zgd_t *zgd, int error)
+	 * so check it explicitly here.
+	 */
+	if (zfs_is_readonly(dstzfsvfs)) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EROFS));
+	}
+
+	/*
+	 * If immutable or not appending then return EPERM.
+	 * Intentionally allow ZFS_READONLY through here.
+	 * See zfs_zaccess_common()
+	 */
+	if ((dstzp->z_pflags & ZFS_IMMUTABLE)) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (SET_ERROR(EPERM));
+	}
+
+	/*
+	 * XXX: There should be no need to sync the source file, as we return
+	 *      an error if somebody is trying to clone a block that was
+	 *      modified in the same transaction group.
+	 */
+#ifdef FRSYNC
+	/*
+	 * If we're in FRSYNC mode, sync out this znode before reading it.
+	 * Only do this for non-snapshots.
+	 *
+	 * Some platforms do not support FRSYNC and instead map it
+	 * to O_SYNC, which results in unnecessary calls to zil_commit. We
+	 * only honor FRSYNC requests on platforms which support it.
+	 */
+	frsync = !!(srcioflag & FRSYNC);
+#endif
+	if (srczfsvfs->z_log &&
+	    (frsync || srczfsvfs->z_os->os_sync == ZFS_SYNC_ALWAYS)) {
+		zil_commit(srczfsvfs->z_log, srczp->z_id);
+	}
+
+	/*
+	 * No overlapping if we are cloning within the same file.
+	 */
+	if (srczp == dstzp) {
+		if ((srcoffset >= dstoffset &&
+		    srcoffset < dstoffset + length) ||
+		    (dstoffset >= srcoffset &&
+		    dstoffset < srcoffset + length)) {
+			zfs_exit_two(srczfsvfs, dstzfsvfs);
+			return (SET_ERROR(EINVAL));
+		}
+	}
+
+	/*
+	 * Maintain predictable lock order.
+	 */
+	if (srczp < dstzp) {
+		srclr = zfs_rangelock_enter(&srczp->z_rangelock, srcoffset,
+		    length, RL_READER);
+		dstlr = zfs_rangelock_enter(&dstzp->z_rangelock, dstoffset,
+		    length, RL_WRITER);
+	} else {
+		dstlr = zfs_rangelock_enter(&dstzp->z_rangelock, dstoffset,
+		    length, RL_WRITER);
+		srclr = zfs_rangelock_enter(&srczp->z_rangelock, srcoffset,
+		    length, RL_READER);
+	}
+
+	srcblksz = srczp->z_blksz;
```

I feel somewhere here must be a check that destination block size is equal to source or the destination file consist of only one block or it will be overwritten completely.  zfs_grow_blocksize() used below will silently do nothing otherwise, that will probably result in data corruption.